pkg/runtime/dispatcher.go

```diff
@@ -103,43 +114,61 @@ type dispatcher struct {
 	// It is not directly used by dispatcher.
 	mapper expr.Evaluator
 
-	// <resolver>
-	resolver atomic.Value
-
 	// gp is used to dispatch multiple adapters concurrently.
 	gp *pool.GoroutinePool
+
+	// protects resolver
+	sync.RWMutex
```

Comment:
Make this a struct field, don't inline it. Making the struct itself a mutex is not what we need: you want a private mutex to guard a field in some methods, but instead you have a public mutex useable by anyone with a `dispatcher` in hand (i.e. the interface this struct presents does not match its intended use, and allows mis-use for no gain).